### PR TITLE
Markov-switching autoregressive models

### PR DESCRIPTION
**Pull request by j-bryan**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
#2160

##### What are the significant changes in functionality due to this change request?
- Adds Markov-switching autoregressive models to the TSA module
- Adds several additional data detrending and transformation algorithms, such as STL decomposition, n-th order differencing, and input CDF preservation (like "preserveInputCDF" option from the ARMA ROM)

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Comment by j-bryan** [comment]
This pull request is being made to record the work I've done on adding Markov-switching autoregressive (MSAR) models to the TSA module as an alternative to ARMA models for noise generation. While assessing these MSAR models, a number of data transformations that seemed useful but that weren't present in TSA were also implemented and are included in this pull request.

At a minimum, the following still need to be added before merging:
- Addition of new classes to documentation
- Regression test that includes sampling from a `MarkovAR` model. The current implementation of `MarkovAR` does not reseed correctly, making it difficult to create reproducible test values.
- Revision of how the numpy RNG engine is used for generating with a `MarkovAR` model to also be compatible with the Crow RNG engine.

File: ravenframework/TSA/Transformers/Normalizers.py [hunk 1]
```diff
@@ -20,9 +20,13 @@
 """
 
 import sklearn.preprocessing as skl
+import numpy as np
+from scipy.stats import iqr
+from copy import deepcopy
```
Comment (dylanjm) [inline ravenframework/TSA/Transformers/Normalizers.py:25]: Are these used for anything?

Comment (j-bryan) [reply ravenframework/TSA/Transformers/Normalizers.py:25]: Leftovers from before I split some transformers off to a different file. I'll remove those.

File: ravenframework/TSA/STL.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+@author: j-bryan
+
+Detrending with STL decomposition
+"""
+
+import numpy as np
+
+from .TimeSeriesAnalyzer import TimeSeriesTransformer, TimeSeriesGenerator
+from ..utils import importerUtils, InputTypes, InputData, xmlUtils
+statsmodels = importerUtils.importModuleLazy('statsmodels', globals())
```
[the remark this answers is not in the document]
Comment (j-bryan) [reply ravenframework/TSA/STL.py:26]: I wasn't sure why exactly it was done this way, but statsmodels is imported like this in the `ARMA` class as well. I followed the pattern because I saw something about using lazy importing of `statsmodels` (and a few other libraries) as standard for RAVEN. Do you know what's going on there that's not accomplished by an import in the `fit` method?

File: ravenframework/TSA/MarkovAR.py [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+      @ In, settings, dict, additional settings specific to algorithm
+      @ Out, residual, np.array, reduced signal shaped [pivotValues, targets]
+    """
+    residual = initial.copy()
+    for tg, (target, data) in enumerate(params.items()):
+      # The model fit results contain the fit residuals, so we can just use those.
+      # The length of the residuals array is shorter than the length of the original signal by the
+      # AR order (not really sure why), so we need to pad the end of the residual array with zeros.
+      modelResid = data['MarkovAR']['model'].resid
+      residual[:, tg] = np.r_[modelResid, np.zeros(len(residual) - len(modelResid))]
```
[the remark this answers is not in the document]
Comment (j-bryan) [reply ravenframework/TSA/MarkovAR.py:297]: Yeah, the residual should be zero mean, but of course there's no guarantee of that. I initially had it as `np.nan` like a masked value, but I didn't want that causing problems downstream for no real reason. I'd say to just leave it, but I'll add a comment in the code that it might be worth revisiting, especially if these residuals are used by another algorithm in the future.

**Review by dylanjm — APPROVED** [review]
Changes LGTM and satisfy merging requirements.
